csds-client/client/v3/client.go

```diff
@@ -72,6 +74,15 @@ func (c *ClientV3) parseNodeMatcher() error {
 		} else if len(networkNameValue) > 0 && len(meshScopeValue) > 0 {
 			return fmt.Errorf("cannot set both %v or %v", gcpNetworkNameKey, gcpMeshScopeKey)
 		}
+
+		meshOrNetworkName := meshScopeValue
+		if len(meshScopeValue) == 0 {
+			meshOrNetworkName = networkNameValue
+		}
+		// node.id is expected to be in the format projects/<project_id>/networks/<mesh/network_name>/nodes/<node_id>
+		// for IAM permissions. For CSDS V3 requests node_id part is randomly generated since the users aren't expected
+		// to pass a node_id.
```

Comment:
I don't really understand why `node.id` needs to be in this format. I'm assuming it's because how Traffic-Director works, and if this client only works with Traffic-Director, then sure, it can be set like that.
If it's supposed to be more generic, I suggest separating the Traffic-Director only code somewhere else.

Reply:
If the case is that only Traffic-Director is supported, then I suggest modifying the name to TD-client or something similar, so it will be clear the the tool doesn't "currently only support GCP's Traffic Director", but it should be a tool that is made to "support only GCP's Traffic Director".

Reply:
Sorry for no activity on this for a while. I confirmed that we should keep the tool generic and not force the `node.id` to be of what TD expects. I propose two solutions here:
1. Enhance this current PR to check if request is for TD then generate node.id of format `projects/<project_id>/networks/<mesh/network_name>/nodes/<uuid>`. Else just pass UUID
2. Revert the current change and continue to read node.if from request.yaml. If CSDS request is for TD, then verify that passed node.id is of `projects/<project_id>/networks/<mesh/network_name>/nodes/<node_id>` format and if not, throw an error.

I am leaning towards option-2. @adisuissa @fuqianggao let me know your thoughts.

Reply:
I think we want option 1. Option 2 won't work, because the node_id in the request.yaml is for the xDS client that you are interested in knowing the detailed config, shouldn't use it for csds client node id.

Reply:
Just to clarify, we have two sets of node id in CSDS V3. 1. One inside the node_matchers which is used to specify the id of the xDS client you are interested in knowing the details like you mentioned. This is same as CSDS V2 and hasn't changed. 2. Other is the [node.id](https://github.com/envoyproxy/envoy/blob/294324a8cb3e94dade9e8b9762cbadab94101c77/api/envoy/service/status/v3/csds.proto#L87) directly inside _ClientStatusRequest_ used to identify where the CSDS request came from. I am referring to this node.id in Option-2. For TD we can pass a dummy id for this to pass IAM check but it has to be in the cloud resource format: projects/<project_id>/networks/<mesh/network_name>/nodes/<uuid>.

Reply:
OK, in this case, I'd prefer vote for option 1.

Reply:
Sorry... I'm not very familiar with this code.
I've looked again at this whole function, and I see that this is part of the `case "gcp":` switch case. If having "gcp" == "Traffic-Director", then I think your current change is good.

If not, then I think option 1 is better, as it decouples the config from what is actually sent in a field and ensures correctness.

Reply:
Good point. The logic under case `"gcp":` is indeed TD specific logic which mandates project number and one of network name or mesh name.

Reply:
@htuch PTAL and merge the change.

Reply:
Thanks @manish-annappa!
LGTM.